File: frontend_go/courses_gui/courses_gui.go
package courses_gui

import (
	"fmt"

	"github.com/pwiecz/go-fltk"
)

const (
	nrows = 5
	ncols = 6
)

type pair struct {
	pos  int
	size int
}

var (
	rows          []pair
	cols          []pair
	CourseWidgets map[string]fltk.Widget
)

const (
	w0 = 1000
	h0 = 700
)

type EditForm struct {
	Widget  *fltk.Grid
	Entries map[string]fltk.Widget
}

func NewEditForm(lines int) EditForm {
	editPanel := fltk.NewGrid(0, 0, 300, 0)
	editPanel.SetLayout(lines, 2, 3, 3)
	editPanel.SetBox(fltk.BORDER_FRAME)
	//	editPanel.End()
	return EditForm{editPanel, map[string]fltk.Widget{}}
}

func (editform EditForm) AddTextline(name string, label string) {
	entry := fltk.NewButton(0, 0, 0, 0)
	entry.SetAlign(fltk.ALIGN_LEFT | fltk.ALIGN_INSIDE)
	entry.SetBox(fltk.PLASTIC_DOWN_BOX)
	entry.SetColor(fltk.ColorFromRgb(255, 255, 200))
	entry.SetCallback(func() { fmt.Println("Clicked") })
	editform.Widget.SetWidget(entry, 0, 1, fltk.GridFill)
	editform.Widget.SetWidget(
		fltk.NewBox(fltk.NO_BOX, 0, 0, 0, 0, label),
		0, 0, fltk.GridFill)
	editform.Entries[name] = entry
}

func Ui() {
	CourseWidgets = map[string]fltk.Widget{}
	win := fltk.NewWindow(w0, h0)

	//

	courses_box := fltk.NewGrid(0, 0, win.W(), win.H())
	courses_box.SetLayout(2, 2, 3, 3)
	courses_box.SetColor(fltk.WHITE)

	topPanel := fltk.NewFlex(0, 0, 0, 100)
	topPanel.SetType(fltk.COLUMN)
	fltk.NewBox(fltk.FLAT_BOX, 0, 0, 0, 0, "TODO")
	panelBox := fltk.NewFlex(0, 0, 0, 0)
	panelBox.SetType(fltk.ROW)
	panelBox.SetMargin(5, 5)

	tableType := fltk.NewChoice(0, 0, 0, 0)
	tableType.ClearVisibleFocus()
	panelBox.Fixed(tableType, 100)
	tableType.Add("Class", nil)
	tableType.Add("Teacher", nil)
	tableType.Add("Subject", nil)
	CourseWidgets["TableType"] = tableType

	tableRow := fltk.NewChoice(0, 0, 0, 0)
	tableRow.ClearVisibleFocus()
	panelBox.Fixed(tableRow, 150)
	tableRow.Add("FB: Fritz Jolander Jeremias Braun", nil)
	tableRow.Add("DG: Diego Garcia", nil)
	tableRow.Add("PM: Pamela Masterson", nil)
	CourseWidgets["TableRow"] = tableRow

	// The label of the output widget allocates no space of its own,
	// overwriting anything to the left of the widget, so add padding.
	pad := fltk.NewBox(fltk.NO_BOX, 0, 0, 0, 0)
	tableTotals := fltk.NewOutput(0, 0, 0, 0)
	tableTotals.SetColor(fltk.ColorFromRgb(255, 255, 200))
	//tableTotals.SetCallback(func() { fmt.Println("Hello there!") })
	tableTotals.SetValue("Read only ˝Öößŋħĸ€")
	tableTotals.ClearVisibleFocus() // no cursor, but text cannot be copied
	tableTotals.SetLabel("Total lessons:")
	w, _ := tableTotals.MeasureLabel()
	panelBox.Fixed(pad, w+20)
	CourseWidgets["TableTotals"] = tableTotals

	panelBox.End()
	topPanel.Fixed(panelBox, 40)

	topPanel.End()
	courses_box.SetWidgetWithSpan(topPanel, 0, 0, 1, 2, fltk.GridFill)

	table := fltk.NewTableRow(0, 0, 0, 0)
	table.EnableColumnHeaders()
	table.DisableRowHeaders()
	table.SetColumnCount(ncols)
	table.SetRowCount(nrows)

	table.SetColumnWidthAll(200)
	table.SetRowHeightAll(50)
	table.SetColumnHeaderHeight(40)
	table.SetType(fltk.SelectSingle)

	table.SetDrawCellCallback(func(tc fltk.TableContext, row, col, x, y, w, h int) {
		if tc == fltk.ContextCell {
			fltk.SetDrawFont(fltk.HELVETICA, 14)
			fltk.DrawBox(fltk.FLAT_BOX, x, y, w, h, fltk.BLACK)
			bg := fltk.WHITE
			if table.IsRowSelected(row) {
				bg = fltk.YELLOW
			}
			fltk.DrawBox(fltk.FLAT_BOX, x+1, y+1, w-2, h-2, bg)
			fltk.SetDrawColor(fltk.BLACK)
			fltk.Draw(fmt.Sprintf("%d", row+col), x, y, w, h, fltk.ALIGN_CENTER)
		}
		if tc == fltk.ContextRowHeader {
			fltk.SetDrawFont(fltk.HELVETICA_BOLD, 14)
			fltk.DrawBox(fltk.UP_BOX, x, y, w, h, fltk.BACKGROUND_COLOR)
			fltk.SetDrawColor(fltk.BLACK)
			fltk.Draw(fmt.Sprintf("row %d", row+1), x, y, w, h, fltk.ALIGN_CENTER)
		}
		if tc == fltk.ContextColHeader {
			fltk.SetDrawFont(fltk.HELVETICA_BOLD, 14)
			fltk.DrawBox(fltk.UP_BOX, x, y, w, h, fltk.BACKGROUND_COLOR)
			fltk.SetDrawColor(fltk.BLACK)
			fltk.Draw(fmt.Sprintf("col %d", col+1), x, y, w, h, fltk.ALIGN_CENTER)
		}
	})

	table.End()
	//		win.Resizable(table)

	courses_box.SetWidget(table, 1, 0, fltk.GridFill)

	// The editing form (right panel)
	editPanel := NewEditForm(6)
	editPanel.AddTextline("CourseBlock", "Course Block")

	/*
		editPanel := fltk.NewGrid(0, 0, 300, 0, "Edit Panel")
		editPanel.SetLayout(6, 2, 3, 3)
		editPanel.SetBox(fltk.BORDER_FRAME)
		//editPanel.SetColor(fltk.WHITE)

		courseBlock := fltk.NewButton(0, 0, 0, 0, "A course")
		courseBlock.SetAlign(fltk.ALIGN_LEFT | fltk.ALIGN_INSIDE)
		courseBlock.SetBox(fltk.PLASTIC_DOWN_BOX)
		courseBlock.SetColor(fltk.ColorFromRgb(255, 255, 200))
		courseBlock.SetCallback(func() { fmt.Println("Clicked") })
		editPanel.SetWidget(courseBlock, 0, 1, fltk.GridFill)
		editPanel.SetWidget(
			fltk.NewBox(fltk.NO_BOX, 0, 0, 0, 0, "Course Block"),
			0, 0, fltk.GridFill)
		editPanel.End()
	*/
	editPanel.Widget.End()
	//editPanel.Entries["CourseBlock"].(*fltk.Button).Deactivate()
	editPanel.Entries["CourseBlock"].(*fltk.Button).SetLabel("This entry is really rather long")

	courses_box.SetWidget(editPanel.Widget, 1, 1, fltk.GridFill)

	//b2 := fltk.NewBox(fltk.BORDER_BOX, 0, 0, 300, 0, "Right Panel")
	//courses_box.SetWidget(b2, 1, 1, fltk.GridFill)

	courses_box.SetColumnWeight(1, 0)
	courses_box.SetRowWeight(0, 0)

	courses_box.End()

	//

	win.Resizable(courses_box)

	win.End()
	win.Show()
	fltk.Run()
}

File: frontend/courses_gui/courses/courses_gui.cpp
#include "courses_gui.h"
#include "editform.h"
#include <FL/Fl.H>
#include <FL/Fl_Choice.H>
#include <FL/Fl_Output.H>
#include <FL/fl_draw.H>

#include <algorithm>
#include <iostream>

CoursesGui::CoursesGui()
    : Fl_Flex(Fl_Flex::COLUMN)
{
    gap(3);
    color(FL_YELLOW);

    // *** Top Panel ***

    // Top Panel – whatever is still needed ...
    auto todo = new Fl_Box(FL_FLAT_BOX, 0, 0, 0, 0, "TODO");
    fixed(todo, 50);

    // Top Panel – the selectors and totals info, at the panel bottom
    auto panelBox = new Fl_Flex(Fl_Flex::ROW);
    panelBox->margin(5, 5);
    fixed(panelBox, 40);

    auto tableType = new Fl_Choice(0, 0, 0, 0);
    tableType->clear_visible_focus();
    panelBox->fixed(tableType, 100);
    tableType->add("Class");
    tableType->add("Teacher");
    tableType->add("Subject");
    Widgets["TableType"] = tableType;

    auto tableRow = new Fl_Choice(0, 0, 0, 0);
    tableRow->clear_visible_focus();
    panelBox->fixed(tableRow, 150);
    tableRow->add("FB: Fritz Jolander Jeremias Braun");
    tableRow->add("DG: Diego Garcia");
    tableRow->add("PM: Pamela Masterson");
    Widgets["TableRow"] = tableRow;

    // The label of the output widget allocates no space of its own,
    // overwriting anything to the left of the widget, so add padding.
    auto pad = new Fl_Box(FL_NO_BOX, 0, 0, 0, 0);
    auto tableTotals = new Fl_Output(0, 0, 0, 0);
    tableTotals->color(0xFFFFC800);
    //tableTotals.SetCallback(func() { fmt.Println("Hello there!") })
    tableTotals->value("Read only ˝Öößŋħĸ€");
    tableTotals->clear_visible_focus(); // no cursor, but text cannot be copied
    tableTotals->label("Total lessons:");
    int wl, hl;
    tableTotals->measure_label(wl, hl);
    panelBox->fixed(pad, wl + 20);
    Widgets["TableTotals"] = tableTotals;

    panelBox->end();

    // Top Panel – end

    // *** The course list/table and course editor ***
    auto mainview = new Fl_Flex(Fl_Flex::ROW);
    mainview->gap(3);
    //mainview->margin(5, 5);

    auto table = new CourseTable();
    Widgets["Table"] = table;

    // *** The course editor form ***
    //TODO ...
    auto editpanel = new EditForm();
    current(NULL);
    editpanel->add_value("CourseBlock", "Course Block");   //->value("");
    editpanel->add_value("BlockSubject", "Block Subject"); //->deactivate();
    //editpanel->add_value("Block Subject")->hide();

    editpanel->add_separator();
    editpanel->add_value("Subject", "Subject");
    editpanel->add_value("Teachers", "Teachers");
    editpanel->add_value("Rooms", "Rooms");
    editpanel->add_value("Units", "Units");
    editpanel->add_list("Constraints", "Constraints");

    // "Properties"

    editpanel->do_layout();

    editpanel->entries[1].widget->deactivate(); //->hide();

    // End of course editor form

    // End of mainview
    mainview->fixed(editpanel, 300);
    mainview->end();

    // End of CoursesGui
    end();
}

void CourseTable::_row_cb(
    void* table)
{
    //TODO
    std::cout << "§§§ " << ((CourseTable*) table)->_current_row << std::endl;
}

CourseTable::CourseTable()
    : Fl_Table_Row(0, 0, 0, 0)
{
    // Rows
    rows(5);            // how many rows
    row_header(0);      // enable row headers (along left)
    row_height_all(30); // default height of rows
    row_resize(0);      // disable row resizing
    // Cols
    cols(6);               // how many columns
    col_header(1);         // enable column headers (along top)
    col_header_height(30); // enable column headers (along top)
    //col_width_all(150);    // default width of columns
    col_resize(0); // disable column resizing
    col_header_color(fl_rgb_color(230, 230, 255));
    type(Fl_Table_Row::SELECT_SINGLE);
    end(); // end the Fl_Table group

    // Temporary (test) data
    headers = {"Subject", "Groups", "Teachers", "Rooms", "Units", "Properties"};
    data = {{"Fr", "11.A", "PM", "fs1", "1, 1, 1", ""},
            {"Fr", "11.B", "PM", "fs1", "1, 1, 1", ""},
            {"Ma", "11", "AH", "k11", "3, 3", "HU"},
            {"Ma", "11.A", "AH", "k11", "1, 1", ""},
            {"Fr", "11.B", "AH", "k11", "1, 1", ""}};
}

void CourseTable::draw_cell(
    TableContext context, int ROW, int COL, int X, int Y, int W, int H)
{
    static char s[40];
    switch (context) {
    case CONTEXT_STARTPAGE: // before page is drawn..
        //fl_font(FL_HELVETICA, 16); // set the font for our drawing operations

        // Adjust column widths
        size_columns();

        // Handle change of selected row
        if (Fl_Table::select_row != _current_row) {
            select_row(Fl_Table::select_row);
            _current_row = Fl_Table::select_row;
            Fl::add_timeout(0.0, _row_cb, this);
        }
        return;
    case CONTEXT_COL_HEADER: // Draw column headers
        fl_push_clip(X, Y, W, H);
        fl_draw_box(FL_THIN_UP_BOX, X, Y, W, H, col_header_color());
        fl_color(FL_BLACK);
        fl_draw(headers[COL].c_str(), X, Y, W, H, FL_ALIGN_CENTER);
        fl_pop_clip();
        return;
    //case CONTEXT_ROW_HEADER:      // Draw row headers
    //    return;
    case CONTEXT_CELL: // Draw data in cells
        fl_push_clip(X, Y, W, H);
        // Draw cell bg
        if (row_selected(ROW)) {
            //if (ROW == _current_row) {
            fl_color(FL_YELLOW);
        } else {
            fl_color(FL_WHITE);
        }
        fl_rectf(X, Y, W, H);
        // Draw cell data
        fl_color(FL_GRAY0);
        fl_draw(data[ROW][COL].c_str(), X, Y, W, H, FL_ALIGN_CENTER);
        // Draw box border
        fl_color(color());
        fl_rect(X, Y, W, H);
        fl_pop_clip();
        return;
    default:
        return;
    }
}

//TODO: Do I need to set font (fl_font()) before using fl_measure()?
void CourseTable::size_columns()
{
    struct colwidth
    {
        int col, width;
    };
    std::vector<colwidth> colwidths;

    int ncols = cols();
    int nrows = rows();
    int w, h, wmax;
    for (int c = 0; c < ncols; ++c) {
        w = 0;
        fl_measure(headers[c].c_str(), w, h, 0);
        wmax = w;
        for (int r = 0; r < nrows; ++r) {
            w = 0;
            fl_measure(data[r][c].c_str(), w, h, 0);
            if (w > wmax)
                wmax = w;
        }
        colwidths.push_back(colwidth{c, wmax});
    }
    std::sort(colwidths.begin(), colwidths.end(), [](colwidth a, colwidth b) {
        return a.width < b.width;
    });

    //for (auto i : colwidths)
    //    std::cout << "$ " << i.col << ": " << i.width << std::endl;

    int restwid = wiw;
    int cols = ncols;
    const int padwidth = 4;
    for (colwidth cw : colwidths) {
        if (cols == 1) {
            if (cw.width + padwidth < restwid) {
                col_width(cw.col, restwid);
            } else {
                col_width(cw.col, cw.width + padwidth);
            }
        } else {
            int defwid = restwid / cols;
            --cols;
            if (cw.width + padwidth < defwid) {
                col_width(cw.col, defwid);
                restwid -= defwid;
            } else {
                col_width(cw.col, cw.width + padwidth);
                restwid -= cw.width + padwidth;
            }
        }
    }
}
